src/sentry/api/serializers/models/dashboard.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import absolute_import
+
+import six
+from sentry.api.serializers import Serializer, register, serialize
+from sentry.models import Dashboard, WidgetDataSource, WidgetDisplayTypes
+
+
+class WidgetSerializer(Serializer):
+    def serialize(self, obj, attrs, user, *args, **kwargs):
+        return {
+            'id': six.text_type(obj.id),
+            'order': six.text_type(obj.order),
+            'title': obj.title,
+            'displayType': WidgetDisplayTypes.get_type_name(obj.display_type),
+            'displayOptions': obj.display_options,
+            'dateAdded': obj.date_added,
+        }
+
+
+class WidgetDataSourceSerializer(Serializer):
+    def serialize(self, obj, attrs, user, *args, **kwargs):
+        return {
+            'id': six.text_type(obj.id),
+            'type': obj.type,
+            'name': obj.name,
+            'data': obj.data,
+            'order': six.text_type(obj.order),
+        }
+
+
+@register(Dashboard)
+class DashboardWithWidgetsSerializer(Serializer):
+
+    def get_attrs(self, item_list, user):
+        widget_data_sources = WidgetDataSource.objects.filter(
+            widget__dashboard_id__in=[i.id for i in item_list]
+        ).select_related('widget')
+        widgets = {}
```

Comment:
Should we have the `WidgetSerializer` handle the data sources? So we could fetch all `Widget`s here, then just serialize them all as a list, and then in `WidgetSerializer.get_attrs`  we handle fetching the data sources for all the widgets. Seems to match the data model more closely.

Reply:
I understand what you're saying, but I can do it in one database call here rather than many several smaller ones. I don't think we should break it up that way because it'll be much slower.

Reply:
It's not really many smaller ones. You need one call to fetch all the widgets, and another to fetch all data sources.